### PR TITLE
Switch from `codespan-reporting` to a custom crate based on codespan

### LINKED ISSUE #29: Consider switching from codespan-reporting to annotate-snippets
# Motivation 

Codespan-reporting is amazing in terms of its files interface, however, it has no middle-level representation, and this has proven to be very very limiting. RSLint could greatly greatly benefit from things such as multiline suggestions like rustc. I think for the future it would be beneficial to make an `rslint_errors` crate which uses annotate-snippets as a backend for rendering, or maybe fork annotate-snippets. 

# Switching 

The change would greatly affect rslint_lexer because it relies on codespan's diagnostic builder, this should be easy to switch out though. Moreover, the file walker would not have to have codespan specific files stuff, it would only need a way of getting the source code for annotate-snippets. Perhaps we should make a Files trait in rslint_errors. Then it should be simple from there. However, diagnostic builders may not work well since rslint_errors cannot rely on rslint_parser since it would be a cyclic dep, therefore we cannot make a Span trait like rslint_core. So we need to either give up the Span trait or make a wrapper of the builder in rslint_core.

### PR DESCRIPTION
**Pull request by Stupremee**
This PR replaces `codespan-reporting` with `annotate-snippets`.
To actually use `annotate-snippets` in a nice manner and to make the switch easier,
a new module, `rslint_errors`, is introduced to provide an API similair to `codespan-reporting`
but build on top of `annotate-snippets`.

Resolves #29

### CODE REVIEW

**Comment by RDambrosio016** [comment]
A diagnostic struct with a lifetime is going to cause a myriad of issues, i pretty much rewrote half of the old linter because of it, so i would rather not do that

**Comment by RDambrosio016** [comment]
oops

**Comment by Stupremee** [comment]
What is the sense of the `Applicability` enum in your code in #29?

**Comment by RDambrosio016** [comment]
Its from rustc_errors https://doc.rust-lang.org/nightly/nightly-rustc/rustc_errors/enum.Applicability.html

**Comment by Stupremee** [comment]
Do we really need the `Applicability` enum?

**Comment by RDambrosio016** [comment]
It does not affect error reporting but it does affect the LSP, since it needs to know what suggestions it can offer which dont have placeholders

**Comment by RDambrosio016** [comment]
Yes the LSP can provide autofix options based on the suggestions from the diagnostic, although im not sure how we are going to do this. Id like to have a fixer for rules, but then the parser cant use it. I think we will keep the fixer for rules but have other suggestions to.

**Comment by Stupremee** [comment]
So, I added all the basic structs and representations that now just need methods and logic to report.
I'm currently not sure if we should stay with `SmolStr` everywhere, or if we should use `String`.
I have no idea how much of a speed boost `SmolStr` would give us.

**Comment by RDambrosio016** [comment]
SmolStr wouldnt do much, most messages are > 22 chars

**Comment by Stupremee** [comment]
I think `rslint_errors` is ready to replace codespan now, except for suggestions, I have to fix some lifetime issues first, and I have to provide an immutable `Diagnostic` (or public methods to read the data) so that the LSP and other stuff can actually apply the suggestions. But other than that it should be ready.

**Comment by RDambrosio016** [comment]
I dont really see why the diagnostic's data should not be public

**Comment by Stupremee** [comment]
Yea that works too

**Comment by Stupremee** [comment]
`annotate-snippets` uses lifetimes everywhere. That's sooo annoying. It destroys everything I want to do :cry:

**Comment by RDambrosio016** [comment]
In what places does it cause issues?

**Comment by Stupremee** [comment]
I have a method `convert` which will convert a `Diagnostic` to a `Snippet`, but all the strings used in `Snippet` have to be a `&str` and thus a lifetime is required, so i can't create new strings in the `convert` method and use them in a `Snippet`. But I think a found a fix.

**Comment by RDambrosio016** [comment]
We could always fork annotate-snippets and include it all in rslint_errors 👀

**Comment by Stupremee** [comment]
image
First progress that actually shows diagnostic :tada: 
(We just allocate 18KiB in 130 allocations for one Diagnostic, I hope this is fine)

**Comment by RDambrosio016** [comment]
Wow that looks amazing.
But why is it allocating 18 KiB? 😰

**Comment by Stupremee** [comment]
I have no idea. I will check if it's my code to convert diagnostic (because it's full of `format` and `to_string`) or if it's `annotate-snippets`

**Comment by RDambrosio016** [comment]
Once you get to the lsp, converting diagnostics should be done in the following way:
- the first primary label defines the range of the diagnostic (all diagnostics should have one, any diagnostics which dont will be skipped)
- Any notes get appended to the main text with a newline, e.g. if theres a note that says `help: oh no :(` and a main message of `uh oh`, then it would be:
```
uh oh
help: oh no
```
- Any secondary (non-primary) labels get turned into diagnostic related info, [lsp type docs here](https://docs.rs/lsp-types/0.82.0/lsp_types/struct.DiagnosticRelatedInformation.html)
- Any suggestions will be turned into notes (and follow note rendering) with their main text

Im pretty sure this is how rustc does it

**Comment by RDambrosio016** [comment]
Then when we have indel based suggestions we can also turn those into commands

**Comment by Stupremee** [comment]
image
image
image

(the line number in the second example is fixed)
Here are some examples of the current state. I think now it's ready to replace codespan

**Comment by RDambrosio016** [comment]
Looks great!

File: crates/rslint_errors/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+[package]
+name = "rslint_errors"
+version = "0.1.0"
+authors = ["Justus K <justus.k@protonmail.com>"]
+edition = "2018"
+
+[dependencies]
+annotate-snippets = "0.9.0"
+rslint_core = { path = "../rslint_core" }
```
[the remark this answers is not in the document]
Comment (Stupremee) [reply crates/rslint_errors/Cargo.toml:9]: This will be fixed, don't worry :smile: 

It just didn't work as I like to do it, but I can fix that

**Comment by RDambrosio016** [comment]
You can have rslint_errors rely on rslint_rowan and impl Span for other things to make it more flexible, like syntax nodes and tokens as well as TextRange

**Comment by Stupremee** [comment]
Thank god you've used an extra `ErrorBuilder` in the parser :smile: 
That saves me sooo much work.

**Comment by RDambrosio016** [comment]
The changes seem to work well, some of the rendering is wonky but we can fix that later, i think we can merge it now

**Comment by RDambrosio016** [comment]
I take that back, it seems we are suffering from the same issue as https://github.com/denoland/deno_lint/pull/391 which is a little odd

**Comment by Stupremee** [comment]
What issue are they suffering from?

**Comment by Stupremee** [comment]
The issues are super weird

**Comment by RDambrosio016** [comment]
Are you sure? it still seems to be a bit broke 🤔 
```
error[SyntaxError]: expected `'('` but instead found `length`
  --> tests\main.js:20:7
   |
...
19 |
20 |   for length;
   |       ^^^^^^ unexpected
   |
error[SyntaxError]: Expected an expression, but found none
  --> tests\main.js:26:5
   |
...
25 |   // when jumpSize = √array.length.
26 |   const jumpSize = Math.floor(Math.sqrt(arraySize));
   |     ^^^^^ Expected an expression here
   |
```

**Comment by RDambrosio016** [comment]
Switching from annotate-snippets to codespan has solved all of the issues, essentially the rendering is the same as using codespan on its own but we have a cool new unified builder as well as suggestions, and with all of the nice codespan rendering 🎉

**Pull request merged**